peerstore/peerstore.go

```diff
@@ -0,0 +1,158 @@
+// Package peerstore provides types and interfaces for local storage of address information, 
+// metadata, and public key material about libp2p peers.
+package peerstore
+
+import (
+	"context"
+	"errors"
+	"io"
+	"math"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/peer"
+
+	ic "github.com/libp2p/go-libp2p-core/crypto"
+
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+var ErrNotFound = errors.New("item not found")
+
+var (
+	// AddressTTL is the expiration time of addresses.
+	AddressTTL = time.Hour
+
+	// TempAddrTTL is the ttl used for a short lived address
+	TempAddrTTL = time.Minute * 2
+
+	// ProviderAddrTTL is the TTL of an address we've received from a provider.
+	// This is also a temporary address, but lasts longer. After this expires,
+	// the records we return will require an extra lookup.
+	ProviderAddrTTL = time.Minute * 10
+
+	// RecentlyConnectedAddrTTL is used when we recently connected to a peer.
+	// It means that we are reasonably certain of the peer's address.
+	RecentlyConnectedAddrTTL = time.Minute * 10
+
+	// OwnObservedAddrTTL is used for our own external addresses observed by peers.
+	OwnObservedAddrTTL = time.Minute * 10
+)
+
+// Permanent TTLs (distinct so we can distinguish between them, constant as they
+// are, in fact, permanent)
+const (
+	// PermanentAddrTTL is the ttl for a "permanent address" (e.g. bootstrap nodes).
+	PermanentAddrTTL = math.MaxInt64 - iota
+
+	// ConnectedAddrTTL is the ttl used for the addresses of a peer to whom
+	// we're connected directly. This is basically permanent, as we will
+	// clear them + re-add under a TempAddrTTL after disconnecting.
+	ConnectedAddrTTL
+)
+
+// Peerstore provides a threadsafe store of Peer related
+// information.
+type Peerstore interface {
+	io.Closer
+
+	AddrBook
+	KeyBook
+	PeerMetadata
+	Metrics
+
+	// PeerInfo returns a peer.PeerInfo struct for given peer.ID.
+	// This is a small slice of the information Peerstore has on
+	// that peer, useful to other services.
+	PeerInfo(peer.ID) peer.AddrInfo
+
+	GetProtocols(peer.ID) ([]string, error)
+	AddProtocols(peer.ID, ...string) error
+	SetProtocols(peer.ID, ...string) error
+	SupportsProtocols(peer.ID, ...string) ([]string, error)
```

Comment:
we should make the protocol stuff into its own interface and compose, I am working to split that off in the peerstore.

Reply:
Host.Mux() may have those methods, since it knows about all protocols (via setProtocolHandler)